### ISSUE TITLE
VAPOR Cache Size knob causing more problems than it solves. 

### ISSUE THREAD

**Issue by shaomeng** [NCAR/VAPOR#2702]
VAPOR global settings has a "Cache Size" entry. However, it seems to cause more problems than it solves. 

The root problem it causes is a more complex logic, which in turn causes many development burdens and bugs. Worse, the development burden is non-trivial, as memory management is tricky and error-prone. Also, the bugs it causes are often non-trivial to debug (#2661). 

The convenience it tries to bring is a more strict resource cap. However, the current state of the code is only half way there, since only one model in VAPOR honors that resource cap. At the same time, it appears that this functionality is rarely used by any user or developer, as the bug that triggers #2661 has been there years without being detected.

To summary, VAPOR cache size attempts to address something very few users actually use, and it's only half way there addressing it, and it causes a lot of other troubles. Thus I suggest we remove it.

**Comment by clyne** [comment]
Are you suggesting that we remove the cap, or remove the cache entirely?

If we remove the cache we'd loose an important performance optimization: the ability to avoid expensive I/O, and in the case of the VDC, decoding operations. If you don't think this would have a significant impact, try loading a large, time-varying VDC data set and set the cache size to something small, but large enough to hold a single variable (and coordinates). Then try going back and forth between different time steps, variables, or regions of interest (Geometry). It's quite painful.

If we remove the cap, then enterprise resources that apply a memory limit such as casper become unusable with large data. VAPOR's memory use will grow and grow and at some point the OS will terminate the application. 

Yes, it is unfortunate that not all of the consumers of large amounts of memory (e.g. the renderers) make use of the VAPOR memory manager. We might want to consider addressing that. However, the VAPOR memory manager is designed to satisfy large memory requests; it's not appropriate for small requests such as every time C++ container classes need to grow. Hence, VAPOR's total memory footprint will always be larger than what the VAPOR memory manager allows.

I disagree with the premise that the cache size causes more problems then it solves. We may simply need to improve it, not throw it away.

**Comment by shaomeng** [comment]
Well, my suggest is to do what most applications do, which is to 1) use resources responsibly and 2) abort when the requested resource is not available. By using resources responsibly, I mean the application requests resources that are necessary to perform its task, and releases resources in a timely manner. By abort when resources not available, I mean delegating the global resource management to the OS, so the application either abort or respond very slowly. 

Regarding your concern on I/O, any modern OS would cache the file and avoid repeated I/O *if* system memory is large enough to hold that file. So, an explicit cache for the purpose of saving I/O would have limited effect.

Regarding your concern on enterprise resources, I don't know why VAPOR's memory will continue to grow *if* the application requests and releases memory responsibly. 

To use an example, Flow Renderer would decide that 3 variables are necessary for non-time-varying rendering and 6 variables are necessary for time-varying rendering. Thus, a Flow Renderer would request that much resource, and if the system cannot afford, VAPOR aborts or responds slowly by using swap. On the other hand, it's the responsibility of Flow Renderer to release variables at T-1 step in non-time-varying case, and release variables at T-2 step in time-varying case. There's no chance the application consumes all memory no matter how many time steps you render. 

If temporarily forget that no render is honoring this cache, the data manager isn't doing a good job making use of it too. This is because, as Scott discovered, a cache of 8GB would not hold a dataset of 7GB.

The bottom line is, there is little evidence showing that folks are making use of the knob. On the contrary, the fact that severe bugs exist years before being discovered suggests that this piece of code is rarely used -- rarely solves problems. I'd assess that it's not worth engineering hours to improve it.

**Comment by clyne** [comment]
> Well, my suggest is to do what most applications do, which is to 1) use resources responsibly and 2) abort when the requested resource is not available. By using resources responsibly, I mean the application requests resources that are necessary to perform its task, and releases resources in a timely manner. By abort when resources not available, I mean delegating the global resource management to the OS, so the application either abort or respond very slowly.

Unfortunately, the system memory manager doesn't meet all needs. That is why some applications implement their own. In our case, we have a specialized application that requires performance and the ability to operate on very large data sets. So it doesn't make sense to compare it to "most applications". If you do some googling you can find discussion on user-level memory management. 

> 
> Regarding your concern on I/O, any modern OS would cache the file and avoid repeated I/O _if_ system memory is large enough to hold that file. So, an explicit cache for the purpose of saving I/O would have limited effect.

That certainly helps, but it doesn't address the wavelet decoding aspect.

> 
> Regarding your concern on enterprise resources, I don't know why VAPOR's memory will continue to grow _if_ the application requests and releases memory responsibly.

I don't know what you mean by responsibly. Again, I suggest you try interacting with a large data set when you can't fit multiple variables into memory. Have you tried that yet?

> 
> To use an example, Flow Renderer would decide that 3 variables are necessary for non-time-varying rendering and 6 variables are necessary for time-varying rendering. Thus, a Flow Renderer would request that much resource, and if the system cannot afford, VAPOR aborts or responds slowly by using swap. On the other hand, it's the responsibility of Flow Renderer to release variables at T-1 step in non-time-varying case, and release variables at T-2 step in time-varying case. There's no chance the application consumes all memory no matter how many time steps you render.

With the changes to the default cache size we've made that is exactly what happens.

> 
> If temporarily forget that no render is honoring this cache, the data manager isn't doing a good job making use of it too. This is because, as Scott discovered, a cache of 8GB would not hold a dataset of 7GB.
> 
> The bottom line is, there is little evidence showing that folks are making use of the knob. On the contrary, the fact that severe bugs exist years before being discovered suggests that this piece of code is rarely used -- rarely solves problems. I'd assess that it's not worth engineering hours to improve it.

Few users will need to cap memory, and that is why we made the default unlimited. However, as I said with the Casper case, if you can't control the memory usage when needed the application could be terminated and that is not what we want either.

**Comment by shaomeng** [comment]
Well, I certainly agree that there's value to have an explicit memory management, but I'm questioning the amount of value it provides and the amount of maintenance it will need. 

I would say leave it there if it delivers what it supposed to deliver, but unfortunately, it doesn't in its current shape:
1. Most of renderers don't honor it;
2. Even the DataManager itself cannot fit in a 7GB dataset with 8GB cache allocated, huh.

A thought exercise, which one of the 2 scenarios is worse if Alice were sitting at Casper help desk:

Scenario A:
Bob: Help! My application runs super slow!
Alice: Logs show that your app is using swap a lot. You need to request more memory when submitting jobs.
Bob: Oh right, now it's a lot faster! 

Scenario B:
Bob: Help! My application runs super slow!
Alice: Logs show that your app is using swap a lot. You need to request more memory when submitting jobs.
Bob: No way. I capped the app memory usage at 40GB and I actually requested 45GB to just be safe.
Alice: I don't know what to say...

**Comment by shaomeng** [comment]
After revisiting this issue, I think my original concern can be mitigated by providing a clear indication in the Settings GUI saying that though the cache value is set to X, the actual memory consumption could be much bigger. A change in GUI in this case is also a much easier measure to take than change of code.

**Comment by clyne** [comment]
Sounds good to me
